modules/core/include/opencv2/core/quaternion.hpp

```diff
@@ -0,0 +1,478 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+//
+//
+//                          License Agreement
+//                For Open Source Computer Vision Library
+//
+// Copyright (C) 2020, Huawei Technologies Co., Ltd. All rights reserved.
+// Third party copyrights are property of their respective owners.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//       http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Author: Liangqian Kong <chargerKong@126.com>
+//         Longbu Wang <riskiest@gmail.com>
+
+#include <opencv2/core.hpp>
+#include <iostream>
+namespace cv
+{
+    #define CV_QUAT_EPS 1.e-6
+    /**
+     * Quaternion are a number system that extends the complex numbers.
+     * A quaternion is generally represented in the form:
+     * 		\f[\begin{array}{rcl}
+     * 		q &= w + x\boldsymbol{i} + y\boldsymbol{j} + z\boldsymbol{k}\\
+     *		  &= [w, x, y, z]\\
+     *		  &= [w, \boldsymbol{v}]
+     *		  \end{array}.\f]
+     * A unit quaternion is usually represents rotation, which has the form:
+     * 		\f[q = [cos\theta, sin\theta\cdot u_x, sin\theta\cdot u_y, sin\theta\cdot u_z].\f]
+     *
+     * In fact. A general quaternion could also be represented in the form
+     * 		\f[q = ||q||[cos\theta, sin\theta\cdot u_x, sin\theta\cdot u_y, sin\theta\cdot u_z].\f]
+     * where \f$||q||\f$ represents the norm of \f$q\f$
+     * all functions will be degenerate to unit cases when \f$||q|| = 1\f$
+     */
+    template <typename _Tp>
+    class Quat
+    {
+    using value_type = typename std::enable_if<
+        std::is_same<float, _Tp>::value||
+        std::is_same<double, _Tp>::value,
+        _Tp
+        >::type;
+    public:
+        Quat() = default;
+        explicit Quat(const cv::Vec<_Tp, 4> &coeff);
+        Quat(_Tp w, _Tp x, _Tp y, _Tp z);
+        Quat(const _Tp angle, const cv::Vec<_Tp, 3> &axis, const _Tp qNorm=1);
+        explicit Quat(const cv::Mat &R);
+        explicit Quat(const Vec<_Tp, 3> &rodrigurs);
```

Comment:
typo Rodrigues

Reply:
Does it make sense to extend cv::Rodruigez? or add cv::RodruigezQ so that it works directly with a 4x1 cv::Mat?

Reply:
Maybe rather cv::RodruigezQ or cv::RodruigezQuat, as currently when a 3x1 vector is put in, a 3x3 matrix is expected.

Reply:
I think the same is OK and easy to understand for users.

Reply:
I meant it would be nice to have a functional way of converting from rVec to qVec (quaternion vector) instead of the objectoriented style with Quat.
cv::rodruigez converts 3x1 mat to a 3x3 mat and vice versa.

Other comment:
Does it make sense to let Quat inherit from cv::Vec<type, 4>?
This way we get all conversions from Quat to cv::Mat for free.

Reply:
It's OK to write the function rvecToQuat outside the class. But we will still keep this constructor, one is for the completeness of the class, and the other is that the functions under the cv outside the class are scattered, which may be difficult for users to find
- - - - - - - - - - - - - - -
I've seen the previous discussion between Rostislav and Vadim.
Rostislav may perfer to make a new type
here is the content
![screenshot](https://user-images.githubusercontent.com/49099366/94359351-3ad6f000-00d9-11eb-88e4-7de1aba85509.png)